### ISSUE TITLE
DDP BERT-Base on SQuaD2.0

### ISSUE THREAD

**Issue by JerryPan2718** [huggingface/transformers#13781]
## Environment info

- `transformers` version: 4.10.2
- Platform: Linux-5.4.0-1056-aws-x86_64-with-Ubuntu-18.04-bionic
- Python version: 3.6.9
- PyTorch version (GPU?): 1.9.1+cu102 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: yes. I used cuda
- Using distributed or parallel set-up in script?: Data parallel

### Who can help

`from argparse import ArgumentParser`
import torch
import torch.distributed as dist
from torch.nn.parallel import DistributedDataParallel as DDP
from torch.utils.data import DataLoader, Dataset
from torch.utils.data.distributed import DistributedSampler
from transformers import DistilBertForQuestionAnswering
from transformers import DistilBertTokenizerFast
import json
from pathlib import Path

SEED = 42
BATCH_SIZE = 8
NUM_EPOCHS = 3

def read_squad(path):
    path = Path(path)
    with open(path, 'rb') as f:
        squad_dict = json.load(f)

    contexts = []
    questions = []
    answers = []
    for group in squad_dict['data']:
        for passage in group['paragraphs']:
            context = passage['context']
            for qa in passage['qas']:
                question = qa['question']
                for answer in qa['answers']:
                    contexts.append(context)
                    questions.append(question)
                    answers.append(answer)

    return contexts, questions, answers

train_contexts, train_questions, train_answers = read_squad('../squad/train-v2.0.json')
val_contexts, val_questions, val_answers = read_squad('../squad/dev-v2.0.json')

def add_end_idx(answers, contexts):
    for answer, context in zip(answers, contexts):
        gold_text = answer['text']
        start_idx = answer['answer_start']
        end_idx = start_idx + len(gold_text)

        # sometimes squad answers are off by a character or two – fix this
        if context[start_idx:end_idx] == gold_text:
            answer['answer_end'] = end_idx
        elif context[start_idx-1:end_idx-1] == gold_text:
            answer['answer_start'] = start_idx - 1
            answer['answer_end'] = end_idx - 1     # When the gold label is off by one character
        elif context[start_idx-2:end_idx-2] == gold_text:
            answer['answer_start'] = start_idx - 2
            answer['answer_end'] = end_idx - 2     # When the gold label is off by two characters

add_end_idx(train_answers, train_contexts)
add_end_idx(val_answers, val_contexts)

tokenizer = DistilBertTokenizerFast.from_pretrained('distilbert-base-uncased')

train_encodings = tokenizer(train_contexts, train_questions, truncation=True, padding=True)
val_encodings = tokenizer(val_contexts, val_questions, truncation=True, padding=True)

def add_token_positions(encodings, answers):
    start_positions = []
    end_positions = []
    for i in range(len(answers)):
        start_positions.append(encodings.char_to_token(i, answers[i]['answer_start']))
        end_positions.append(encodings.char_to_token(i, answers[i]['answer_end'] - 1))

        # if start position is None, the answer passage has been truncated
        if start_positions[-1] is None:
            start_positions[-1] = tokenizer.model_max_length
        if end_positions[-1] is None:
            end_positions[-1] = tokenizer.model_max_length

    encodings.update({'start_positions': start_positions, 'end_positions': end_positions})

add_token_positions(train_encodings, train_answers)
add_token_positions(val_encodings, val_answers)

class SquadDataset(torch.utils.data.Dataset):
    def __init__(self, encodings):
        self.encodings = encodings

    def __getitem__(self, idx):
        return {key: torch.tensor(val[idx]) for key, val in self.encodings.items()}

    def __len__(self):
        return len(self.encodings.input_ids)

train_dataset = SquadDataset(train_encodings)
val_dataset = SquadDataset(val_encodings)

def main():
    parser = ArgumentParser('DDP usage example')
    parser.add_argument('--local_rank', type=int, default=-1, metavar='N', help='Local process rank.')  # you need this argument in your scripts for DDP to work
    args = parser.parse_args()

    # keep track of whether the current process is the `master` process
    args.is_master = args.local_rank == 0
    args.device = torch.cuda.device(args.local_rank)

    torch.cuda.set_device(args.local_rank)

    # set the seed for all GPUs (also make sure to set the seed for random, numpy, etc.)
    torch.cuda.manual_seed_all(SEED)
 
    # initialize your model (BERT in this example)
    model = DistilBertForQuestionAnswering.from_pretrained("distilbert-base-uncased")
    # model = BertForMaskedLM.from_pretrained('bert-base-uncased')
    model = torch.nn.DataParallel(model)
    model = model.to(args.device)

    model = DDP(
        model,
        device_ids=[args.local_rank],
        output_device=args.local_rank
    ) 

    train_dataset = SquadDataset(train_encodings)
    train_sampler = DistributedSampler(train_dataset)
    train_dataloader = DataLoader(train_dataset, sampler=train_sampler, batch_size=batch_size)
    val_dataset = SquadDataset(val_encodings)    
    val_sampler = DistributedSampler(val_dataset)
    val_dataloader = DataLoader(val_dataset, sampler=train_sampler, batch_size=batch_size)   

    for epoch in range(NUM_EPOCHS):
        model.train()
        # let all processes sync up before starting with a new epoch of training
        dist.barrier()
        for step, batch in enumerate(train_dataloader):
            batch = tuple(t.to(args.device) for t in batch)
            outputs = model(*batch)
            loss = outputs[0]

if __name__ == '__main__':
    main()
`

## Information

Model I am using (Bert, XLNet ...): DistilBert

The problem arises when using:
* [ ] the official example scripts: (give details below)
* [x] my own modified scripts: (give details below)

The tasks I am working on is:
* [x] an official GLUE/SQUaD task: (give the name)
* [ ] my own task or dataset: (give details below)

## To reproduce

1. Run the above code on cuda-available devices and get the bug below.

## Expected behavior

<img width="896" alt="Screen Shot 2021-09-28 at 16 15 13" src="https://user-images.githubusercontent.com/37657480/135178052-074ce67c-98e6-4eb8-80c7-7eb494159ed3.png">
